Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs

```diff
@@ -0,0 +1,356 @@
+        {
+            if (solution.TotalVolume == 0)
+            {
+                return false;
+            }
+            var result = _contents.TryAddSolution(solution);
+            if (!result)
+            {
+                return false;
+            }
+
+            UpdateStatus();
+            CheckOverflow();
+            if (checkForEvaporate)
+            {
+                CheckEvaporate();
+            }
+
+            UpdateAppearance();
+            if (!sound)
+            {
+                return true;
+            }
+
+            _entitySystemManager.GetEntitySystem<AudioSystem>().Play(_spillSound);
+            return true;
+        }
+
+        internal Solution SplitSolution(int quantity)
+        {
+            var split = _contents.SplitSolution(quantity);
+            CheckEvaporate();
+            UpdateAppearance();
+            return split;
+        }
+
+        public void CheckEvaporate()
+        {
+            if (CurrentVolume == 0)
+            {
+                Owner.Delete();
+            }
+        }
+
+        private void UpdateStatus()
+        {
+            // If UpdateStatus is getting called again it means more fluid has been updated so let's just wait
+            _evaporationToken?.Cancel();
+
+            if (CurrentVolume > _evaporateThreshold)
+            {
+                return;
+            }
+
+            _evaporationToken = new CancellationTokenSource();
+
+            // KYS to evaporate
+            Timer.Spawn(TimeSpan.FromSeconds(_evaporateTime), CheckEvaporate, _evaporationToken.Token);
+        }
+
+        private void UpdateAppearance()
+        {
+            if (Owner.Deleted)
+            {
+                return;
+            }
+            // Opacity based on level of fullness to overflow
+            // Hard-cap lower bound for visibility reasons
+            var volumeScale = ((float)CurrentVolume / OverflowVolume) * 0.75f + 0.25f;
+            var cappedScale = Math.Min(1.0f, volumeScale);
+            // Color based on the underlying solutioncomponent
+            Color newColor;
+            if (_recolor)
+            {
+                newColor = new Color(
+                    _contents.SubstanceColor.R,
+                    _contents.SubstanceColor.G,
+                    _contents.SubstanceColor.B,
+                    cappedScale);
+            }
+            else
+            {
+                newColor = new Color(
+                    _spriteComponent.Color.R,
+                    _spriteComponent.Color.G,
+                    _spriteComponent.Color.B,
+                    cappedScale);
+            }
+
+            _spriteComponent.Color = newColor;
+
+            _spriteComponent.Dirty();
+
+        }
+
+        /// <summary>
+        /// Will overflow this entity to neighboring entities if required
+        /// </summary>
+        private void CheckOverflow()
```

Comment:
It would be a lot better if this adjacent tile code used the snap grid. You can check icon smoothing as a rough example since those are kind of related.